Add torch.utils._sympy.interp

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #95063
* __->__ #94985
* #94944
* #94939
* #94906

This utility allows us to conveniently abstract interpret Sympy expressions with respect to some alternative domain. I am particularly interested in using ValueRanges to do range analysis on expressions (not this PR).

Some minor house-keeping:
* ReferenceAnalysis got moved to its own file, sprouted a constant() implementation, and some uses of math.* got converted to sympy.*
* ValueRangeAnalysis now understands mod
* Test file gets moved from `test_value_ranges.py` to `test_sympy_utils.py`

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
Add torch.utils._sympy.interp

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add torch.utils._sympy.interp"


Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add torch.utils._sympy.interp"


This utility allows us to conveniently abstract interpret Sympy expressions with respect to some alternative domain. I am particularly interested in using ValueRanges to do range analysis on expressions (not this PR).

Some minor house-keeping:
* ReferenceAnalysis got moved to its own file, sprouted a constant() implementation, and some uses of math.* got converted to sympy.*
* ValueRangeAnalysis now understands mod
* Test file gets moved from `test_value_ranges.py` to `test_sympy_utils.py`

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add torch.utils._sympy.interp"


This utility allows us to conveniently abstract interpret Sympy expressions with respect to some alternative domain. I am particularly interested in using ValueRanges to do range analysis on expressions (not this PR).

Some minor house-keeping:
* ReferenceAnalysis got moved to its own file, sprouted a constant() implementation, and some uses of math.* got converted to sympy.*
* ValueRangeAnalysis now understands mod
* Test file gets moved from `test_value_ranges.py` to `test_sympy_utils.py`

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

Skylion007 commented on .lintrunner.toml:
```diff
@@ -854,6 +854,8 @@ include_patterns = [
     'test/distributed/_composable/**/*.py',
     'torch/testing/_internal/common_dist_composable.py',
     'test/test_value_ranges.py',
+    'torch/utils/_sympy/interp.py',
```
Isn't this a bit extreme just for something that could have been a type ignore comment?

ezyang replied:
this is the UFMT section lol

Skylion007 replied:
Oh lol.

ezyang committed:
Update on "Add torch.utils._sympy.interp"


This utility allows us to conveniently abstract interpret Sympy expressions with respect to some alternative domain. I am particularly interested in using ValueRanges to do range analysis on expressions (not this PR).

Some minor house-keeping:
* ReferenceAnalysis got moved to its own file, sprouted a constant() implementation, and some uses of math.* got converted to sympy.*
* ValueRangeAnalysis now understands mod
* Test file gets moved from `test_value_ranges.py` to `test_sympy_utils.py`

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

eellison commented on torch/utils/_sympy/interp.py:
```diff
@@ -0,0 +1,92 @@
+    # Recursive case
+    args = [sympy_interp(analysis, bindings, arg) for arg in expr.args]  # type: ignore[arg-type]
+    handler = getattr(analysis, handlers()[expr.func])
+    if handler in ASSOCIATIVE_OPS:
```
would we ever call `max(x, y, z)` ? that is surprising to me. I did not realize we have variadic ops here. Or maybe this is due to sympy ?

ezyang replied:
if you write x + y + z you get args = [x, y, z]; sympy will make it n-ary.

```
>>> sympy.Max(x, y, z).args
(x, y, z)
```

eellison commented on torch/utils/_sympy/reference.py:
```diff
@@ -0,0 +1,122 @@
+# The normal Python interpretation of the operators
+# NB: For magic methods this needs to use normal magic methods
+# so that test_magic_methods works
+class ReferenceAnalysis:
```
I think most of these could be more succinct as a mapping from name to operator. E.g. `ge -> operator.ge`.

ezyang replied:
do you actually want them metaprogrammed? It seems easier to understand if we don't

eellison commented on torch/utils/_sympy/reference.py:
```diff
@@ -0,0 +1,122 @@
+# so that test_magic_methods works
+class ReferenceAnalysis:
+    @staticmethod
+    def constant(c, dtype):
```
ignored dtype 🤔

ezyang replied:
that's what ValueRangeAnalysis did too lol

ezyang replied:
I actually don't really understand how this part is supposed to work, if someone could explain it to me that would be great.

eellison replied:
Passes in a constant value and an associated torch dtype: https://github.com/pytorch/pytorch/blob/master/torch/_inductor/lowering.py#L1518

ezyang replied:
This seems... OK to ignore? Because we don't have any notion of size on the reference analysis. Unless I'm obligated to convert an int constant into float if the constant and dtype mismatch?

eellison replied:
Yes, i think we're fine to ignore it. Although it would be more safe to do the conversion in case the callee had a mismatch

eellison reviewed (approved):
Nice! I think we could get rid of a lot of the boiler plate in ReferenceAnalysis with a mapping from name to operator.

ezyang commented on torch/utils/_sympy/value_ranges.py:
```diff
@@ -339,6 +339,14 @@ def log(x):
 
+    @staticmethod
+    def mod(x, y):
+        if x.lower == x.upper and y.lower == y.upper and y.lower != 0:
```
`is_singleton` maybe?

ezyang replied:
this is actually a pretty big codemod, I'll do it in a follow up

ezyang replied:
ok actually it's not that big, folding in

ezyang commented on torch/utils/_sympy/interp.py:
```diff
@@ -0,0 +1,92 @@
+):
+    # Handle base cases
+    # TODO: not really sure if I'm passing the right dtype here
+    # TODO: wouldn't it be better to pass the sympy expression through
+    # sometimes?
```
So when I was testing, I was running everything with sympy expressions to avoid disagreements between float and sympy symbolic. But because constant() only takes int/float, I have to int/float the sympy expression before calling it into constant (and then the reference analysis turns it back into sympy).

ezyang commented on torch/utils/_sympy/interp.py:
```diff
@@ -0,0 +1,92 @@
+ASSOCIATIVE_OPS = {"minimum", "maximum", "mul", "add", "and_", "or_"}
+
+
+def sympy_interp(
```
But you have no idea what the return type is lol. That's analysis dependent.

ezyang committed:
Update on "Add torch.utils._sympy.interp"


This utility allows us to conveniently abstract interpret Sympy expressions with respect to some alternative domain. I am particularly interested in using ValueRanges to do range analysis on expressions (not this PR).

Some minor house-keeping:
* ReferenceAnalysis got moved to its own file, sprouted a constant() implementation, and some uses of math.* got converted to sympy.*
* ValueRangeAnalysis now understands mod
* Test file gets moved from `test_value_ranges.py` to `test_sympy_utils.py`

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add torch.utils._sympy.interp"


This utility allows us to conveniently abstract interpret Sympy expressions with respect to some alternative domain. I am particularly interested in using ValueRanges to do range analysis on expressions (not this PR).

Some minor house-keeping:
* ReferenceAnalysis got moved to its own file, sprouted a constant() implementation, and some uses of math.* got converted to sympy.*
* ValueRangeAnalysis now understands mod
* Test file gets moved from `test_value_ranges.py` to `test_sympy_utils.py`

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge

ezyang commented:
@pytorchbot merge -f "known flakiness"
